File: Laboratory 5/Laboratory_5.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
from skimage import color, filters, measure, segmentation, util
from skimage.filters import threshold_multiotsu, gabor_kernel
from sklearn.cluster import KMeans

# Load image
img = cv2.imread('flower.jpg')  # Replace 'flower.jpg' with the correct image path
img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

# Global Image thresholding using Otsu's method
# Calculate threshold using Otsu's method
_, bw = cv2.threshold(img_gray, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)

# Display the original image and the binary image
plt.figure(figsize=(10, 5))
plt.subplot(1, 2, 1)
plt.imshow(img_gray, cmap='gray')
plt.title('Original Image')
plt.subplot(1, 2, 2)
plt.imshow(bw, cmap='gray')
plt.title('Binary Image')
plt.show()

# Multi-level thresholding using Otsu's method
levels = threshold_multiotsu(img_gray, classes=3)
seg_img = np.digitize(img_gray, bins=levels)

# Display the original image and the segmented image
plt.figure(figsize=(10, 5))
plt.subplot(1, 2, 1)
plt.imshow(img_gray, cmap='gray')
plt.title('Original Image')
plt.subplot(1, 2, 2)
plt.imshow(seg_img, cmap='gray')
plt.title('Segmented Image')
plt.show()

# Global histogram threshold using Otsu's method
# Calculate a 16-bin histogram for the image
counts, x = np.histogram(img_gray, bins=16)
plt.figure()
plt.stem(x[:-1], counts, basefmt=" ")
plt.title('Histogram')
plt.show()

# Compute a global threshold using the histogram counts
T = filters.threshold_otsu(img_gray)
bw = img_gray > T

# Create a binary image using the computed threshold and display the image
plt.figure()
plt.imshow(bw, cmap='gray')
plt.title('Binary Image')
plt.show()

# Region-based segmentation using K means clustering
img2 = cv2.imread('flower.jpg')  # Replace with the correct image path
bw_img2 = cv2.cvtColor(img2, cv2.COLOR_BGR2GRAY)

# K-means clustering
kmeans = KMeans(n_clusters=3, random_state=0).fit(bw_img2.reshape(-1, 1))
L = kmeans.labels_.reshape(bw_img2.shape)
B = color.label2rgb(L, image=bw_img2, bg_label=0)

plt.figure()
plt.imshow(B)
plt.title('Labeled Image')
plt.show()

# Connected-component labeling
bin_img2 = bw_img2 > filters.threshold_otsu(bw_img2)
labeledImage = measure.label(bin_img2)
numberOfComponents = labeledImage.max()

# Display the number of connected components
print(f'Number of connected components: {numberOfComponents}')

# Assign a different color to each connected component
coloredLabels = color.label2rgb(labeledImage, bg_label=0)

# Display the labeled image
plt.figure()
plt.imshow(coloredLabels)
plt.title('Labeled Image')
plt.show()

# Adding noise to the image then segmenting it using Otsu's method
img_noise = util.random_noise(img_gray, mode='s&p', amount=0.09)
img_noise = (img_noise * 255).astype(np.uint8)

# Multi-level thresholding using Otsu's method
levels_noise = threshold_multiotsu(img_noise, classes=3)
seg_img_noise = np.digitize(img_noise, bins=levels_noise)

# Display the original image and the segmented image
plt.figure(figsize=(10, 5))
plt.subplot(1, 2, 1)
plt.imshow(img_noise, cmap='gray')
plt.title('Original Image with Noise')
plt.subplot(1, 2, 2)
plt.imshow(seg_img_noise, cmap='gray')
plt.title('Segmented Image with Noise')
plt.show()

# Segment the image into two regions using k-means clustering
RGB = cv2.imread('flower.jpg')  # Replace with the correct image path
RGB_gray = cv2.cvtColor(RGB, cv2.COLOR_BGR2GRAY)
kmeans = KMeans(n_clusters=2, random_state=0).fit(RGB_gray.reshape(-1, 1))
L = kmeans.labels_.reshape(RGB_gray.shape)
B = color.label2rgb(L, image=RGB_gray, bg_label=0)

plt.figure()
plt.imshow(B)
plt.title('Labeled Image')
plt.show()

# Create a set of 24 Gabor filters, covering 6 wavelengths and 4 orientations
wavelengths = np.array([3, 6, 12, 24, 48, 96])
orientations = np.array([0, 45, 90, 135])
gabor_filters = [gabor_kernel(wavelength, theta=np.deg2rad(orientation))
                 for wavelength in wavelengths for orientation in orientations]

# Filter the grayscale image using the Gabor filters and display the filtered images in a montage
gabormag = [cv2.filter2D(RGB_gray, cv2.CV_8UC3, np.real(gabor_filter)) for gabor_filter in gabor_filters]
gabormag = np.array(gabormag)
gabormag = np.transpose(gabormag, (1, 2, 0))

plt.figure(figsize=(12, 8))
for i in range(len(gabor_filters)):
    plt.subplot(4, 6, i+1)
    plt.imshow(gabormag[:, :, i], cmap='gray')
    plt.axis('off')
plt.suptitle('Filtered Images using Gabor Filters')
plt.show()

# Smooth each filtered image to remove local variations
gabormag_smoothed = np.array([cv2.GaussianBlur(gabormag[:, :, i], (0, 0), sigmaX=3*0.5*wavelengths[i//4]) for i in range(len(gabor_filters))])
gabormag_smoothed = np.transpose(gabormag_smoothed, (1, 2, 0))

plt.figure(figsize=(12, 8))
for i in range(len(gabor_filters)):
    plt.subplot(4, 6, i+1)
    plt.imshow(gabormag_smoothed[:, :, i], cmap='gray')
    plt.axis('off')
plt.suptitle('Smoothed Gabor Filtered Images')
plt.show()

# Get the x and y coordinates of all pixels in the input image
nrows, ncols = RGB_gray.shape
X, Y = np.meshgrid(np.arange(ncols), np.arange(nrows))
feature_set = np.dstack((RGB_gray, gabormag_smoothed, X, Y))

# Segment the image into two regions using k-means clustering with the supplemented feature set
kmeans = KMeans(n_clusters=2, random_state=0).fit(feature_set.reshape(-1, feature_set.shape[2]))
L2 = kmeans.labels_.reshape(RGB_gray.shape)
C = color.label2rgb(L2, image=RGB, bg_label=0)

plt.figure()
plt.imshow(C)
plt.title('Labeled Image with Additional Pixel Information')
plt.show()

File: Laboratory 5/Laboratory_5.m
% Global Image thresholding using Otsu's method
% Load image
img = imread('flower.jpg'); % Replace 'image.jpg' with the correct image path

% Convert image to grayscale if it is not already
if size(img, 3) == 3
    img = rgb2gray(img);
end

% Calculate threshold using graythresh
level = graythresh(img);

% Convert into binary image using the computed threshold
bw = imbinarize(img, level);

% Display the original image and the binary image
figure(1); 
imshowpair(img, bw, 'montage'); 
title('Original Image (left) and Binary Image (right)');

% Multi-level thresholding using Otsu's method
% Calculate multiple thresholds using multithresh
levels = multithresh(img);

% Segment the image into regions using the imquantize function
seg_img = imquantize(img, levels);

% Display the original image and the segmented image
figure(2); 
imshowpair(img, seg_img, 'montage'); 
title('Original Image (left) and Segmented Image (right)');

% Global histogram threshold using Otsu's method
% Calculate a 16-bin histogram for the image
[counts, x] = imhist(img, 16);
figure(3);
stem(x, counts);
title('Histogram');

% Compute a global threshold using the histogram counts
T = otsuthresh(counts);

% Create a binary image using the computed threshold and display the image
bw = imbinarize(img, T);
figure(4); 
imshow(bw); 
title('Binary Image');

% Region-based segmentation using K means clustering
img2 = imread('flower.jpg'); % Replace with the correct image path

% Convert the image to grayscale
bw_img2 = im2gray(img2);
figure(5);
imshow(bw_img2);
title('Grayscale Image');

% Segment the image into three regions using k-means clustering
[L, centers] = imsegkmeans(bw_img2, 3);
B = labeloverlay(bw_img2, L);
figure(6); 
imshow(B); 
title('Labeled Image');

% Connected-component labeling
% Convert the image into binary
bin_img2 = imbinarize(bw_img2);

% Label the connected components
[labeledImage, numberOfComponents] = bwlabel(bin_img2);

% Display the number of connected components
disp(['Number of connected components: ', num2str(numberOfComponents)]);

% Assign a different color to each connected component
coloredLabels = label2rgb(labeledImage, 'hsv', 'k', 'shuffle');

% Display the labeled image
figure(7);
imshow(coloredLabels); 
title('Labeled Image');

% Adding noise to the image then segmenting it using Otsu's method
img_noise = imnoise(img, 'salt & pepper', 0.09);

% Calculate single threshold using multithresh
level_noise = multithresh(img_noise);

% Segment the image into two regions using the imquantize function
seg_img_noise = imquantize(img_noise, level_noise);

% Display the original image and the segmented image
figure(8); 
imshowpair(img_noise, seg_img_noise, 'montage'); 
title('Original Image (left) and Segmented Image with noise (right)');

% Segment the image into two regions using k-means clustering
RGB = imread('flower.jpg'); % Replace with the correct image path
L = imsegkmeans(RGB, 2); 
B = labeloverlay(RGB, L);
figure(9); 
imshow(B); 
title('Labeled Image');

% Create a set of 24 Gabor filters, covering 6 wavelengths and 4 orientations
wavelength = 2.^(0:5) * 3;
orientation = 0:45:135;
g = gabor(wavelength, orientation);

% Convert the image to grayscale
bw_RGB = im2gray(im2single(RGB));

% Filter the grayscale image using the Gabor filters and display the filtered images in a montage
gabormag = imgaborfilt(bw_RGB, g); 
figure(10); 
montage(gabormag, "Size", [4 6]);

% Smooth each filtered image to remove local variations
for i = 1:length(g)
    sigma = 0.5 * g(i).Wavelength;
    gabormag(:,:,i) = imgaussfilt(gabormag(:,:,i), 3 * sigma);
end
figure(11); 
montage(gabormag, "Size", [4 6]);

% Get the x and y coordinates of all pixels in the input image
nrows = size(RGB, 1);
ncols = size(RGB, 2);
[X, Y] = meshgrid(1:ncols, 1:nrows); 
featureSet = cat(3, bw_RGB, gabormag, X, Y);

% Segment the image into two regions using k-means clustering with the supplemented feature set
L2 = imsegkmeans(featureSet, 2, "NormalizeInput", true); 
C = labeloverlay(RGB, L2);
figure(12);
imshow(C); 
title("Labeled Image with Additional Pixel Information");
